{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "68619d50-84ff-406b-91ce-c3683613dc53",
   "metadata": {},
   "source": [
    "# Import packages and data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baa19982-2a34-418d-92b9-637ba5872f51",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from natsort import natsorted\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from sklearn import metrics\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "\n",
    "sns.set(font_scale=1, style=\"ticks\", context=\"paper\", palette=\"tab10\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00ef713b-0f07-43f3-bb09-69b7b54eadb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata=sc.read_h5ad('COVID_dataset_scvi_OSN.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01ff525d-f971-4607-96a1-536eb2d0f2e1",
   "metadata": {},
   "source": [
    "# Get OR info per cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71ee6ae-3174-4ad1-867d-b675b17d615f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add info about human ORs\n",
    "#using annotation from Barnes et al. 2020\n",
    "human_anno = pd.read_excel(\n",
    "    \"https://static-content.springer.com/esm/art%3A10.1186%2Fs12864-020-6583-3/MediaObjects/12864_2020_6583_MOESM2_ESM.xlsx\",\n",
    "    sheet_name=0,\n",
    ")\n",
    "human_anno.columns = human_anno.columns.map(lambda l: l.replace(\" \", \"_\").lower())\n",
    "bm_human = (\n",
    "    human_anno[[\"gene_symbol\", \"gene_name\", \"chromosome\", \"strand\", \"ensembl_gene_id\"]]\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    "    .set_index(\"gene_symbol\")\n",
    ")\n",
    "ct = (\n",
    "    human_anno[human_anno.gene_symbol.isin(bm_human.index)]\n",
    "    .groupby(\"gene_symbol\")\n",
    "    .transcript_biotype.value_counts()\n",
    "    .unstack()\n",
    "    .replace(np.nan, 0)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0a2fc4c-e448-4859-b847-d2af4868d5e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "bm_human = bm_human.join(ct[[\"protein_coding\", \"unprocessed_pseudogene\"]])\n",
    "func_human = bm_human[[\"ensembl_gene_id\"]].reset_index()\n",
    "func_human.columns = [\"gene\", \"Ens\"]\n",
    "is_olfr = adata.var.gene_ids.isin(bm_human.ensembl_gene_id)\n",
    "adata.obs[\"olfr_sum\"] = adata.layers[\"counts\"][:, is_olfr].sum(1).A.flatten()\n",
    "adata.obs[\"olfr_max\"] = adata.layers[\"counts\"][:, is_olfr].max(1).A.flatten()\n",
    "adata.obs[\"olfr_norm_sum\"] = adata.layers[\"norm\"][:, is_olfr].sum(1).A.flatten()\n",
    "adata.obs[\"olfr_norm_max\"] = adata.layers[\"norm\"][:, is_olfr].max(1).A.flatten()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51a785f5-820b-4023-876e-a4501005bf41",
   "metadata": {},
   "outputs": [],
   "source": [
    "OLFR_THRESH = 2\n",
    "olfr_raw = adata.layers[\"counts\"][:, is_olfr].A\n",
    "has_any_olfr = (olfr_raw > OLFR_THRESH).any(1)\n",
    "olfr_names = adata.var_names[is_olfr]\n",
    "df_olfr_family = pd.DataFrame(\n",
    "    [re.split(\"(OR[\\\\d]+)([A-Z])\", s)[1:] for s in olfr_names],\n",
    "    columns=[\"family\", \"sub_family\", \"family_number\"],\n",
    "    index=olfr_names,\n",
    ")\n",
    "df_olfr = pd.DataFrame(\n",
    "    olfr_names[olfr_raw.argmax(1)], index=adata.obs_names, columns=[\"top_Olfr\"]\n",
    ")[has_any_olfr].merge(df_olfr_family, left_on=\"top_Olfr\", right_index=True, how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec8a2f12-be36-4709-abae-1c529487b7f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = adata.obs.copy()\n",
    "cat_cols = df.columns[df.dtypes == \"category\"]\n",
    "df[cat_cols] = df[cat_cols].astype(str)\n",
    "df_umap = pd.DataFrame(\n",
    "    adata.obsm[\"X_umap\"], index=adata.obs_names, columns=[\"UMAP1\", \"UMAP2\"]\n",
    ")\n",
    "df = df.join(df_umap)\n",
    "df[\"has_OR\"] = has_any_olfr * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84f8043b-9f1d-4c17-8c07-ca5c21ba3392",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Annotations for plotting\n",
    "#For example\n",
    "COVID = \"covid\"\n",
    "CONTROL = \"normosmic\"\n",
    "CONDS = (CONTROL, COVID)\n",
    "N_COVID = \"COVID\"\n",
    "N_CONTROL = \"control\"\n",
    "N_CONDS = (N_CONTROL, N_COVID)\n",
    "N_COLOR_MAP = dict(zip(N_CONDS, plt.cm.Set1.colors))\n",
    "cond_mapping = dict(zip(CONDS, N_CONDS))\n",
    "PATIENTS = [\n",
    "    \"COVID_1\",\n",
    "    \"COVID_2\",\n",
    "    \"Normosmic_1\",\n",
    "    \"Normosmic_2\"\n",
    "]\n",
    "is_patient = df.orig_ident.isin(PATIENTS)\n",
    "df[\"cond2\"] = df.cond.map(cond_mapping)\n",
    "df_patient = df[is_patient].copy()\n",
    "df_anno = (\n",
    "    df.groupby([\"cond\", \"orig_patients\"], as_index=False)\n",
    "    .total_counts.count()\n",
    "    .set_index(\"orig_patients\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1821bcd-17cb-48d6-83ed-e4f5478dee38",
   "metadata": {},
   "source": [
    "# Make plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "823c83cf-3bc3-4bd4-8400-bc6ae05bef54",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot OR expressing cells\n",
    "has_OR = df_patient[\"has_OR\"] > 0\n",
    "xu, yu = df_patient[[\"UMAP1\", \"UMAP2\"]].values.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b436daec-740b-4429-ab87-f6782daa77d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(1.5, 1.5))\n",
    "ax.scatter(xu[~has_OR], yu[~has_OR], lw=0, s=3, color=\"0.5\", alpha=0.5, label=\"no OR\")\n",
    "ax.scatter(\n",
    "    xu[has_OR], yu[has_OR], lw=0, s=3, color=\"k\", alpha=0.85, label=\"OR-expressing\"\n",
    ")\n",
    "ax.axis(\"off\")\n",
    "leg = ax.legend(\n",
    "    frameon=True,\n",
    "    loc=\"lower center\",\n",
    "    bbox_to_anchor=(0.5, 0.975),\n",
    "    handletextpad=0,\n",
    "    ncol=2,\n",
    "    columnspacing=0,\n",
    "    markerscale=2,\n",
    ")\n",
    "leg.get_frame().set_linewidth(0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c44b8564-a554-4e55-8464-357801797e64",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(1.5, 1.5))\n",
    "\n",
    "sns.barplot(\n",
    "    data=df_patient,\n",
    "    x=\"osn_clusters\",\n",
    "    y=\"has_OR\",\n",
    "    order=order,\n",
    "    hue=\"cond2\",\n",
    "    palette=\"Set1\",\n",
    "    errwidth=0.75,\n",
    ")\n",
    "sns.despine()\n",
    "ax.set_ylim(0, 100)\n",
    "ax.legend(loc=\"upper left\", bbox_to_anchor=(0, 1.05))\n",
    "ax.set_xlabel(\"Cluster\")\n",
    "ax.set_ylabel(\"OR-expressing cells (%)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2df94290-9d20-4782-98f5-a840c91a632d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot OR expression\n",
    "fig, ax = plt.subplots(figsize=(1.5, 1.5))\n",
    "cbar_ax = fig.add_axes([0.95, 0.2, 0.05, 0.6])\n",
    "df_patient[\"olfr_log\"] = np.log1p(df_patient[\"olfr_norm_max\"])\n",
    "c = df_patient[\"olfr_log\"]\n",
    "vmax = np.percentile(c, 99)\n",
    "im = ax.scatter(\n",
    "    xu,\n",
    "    yu,\n",
    "    c=c,\n",
    "    s=3,\n",
    "    cmap=\"cmo.matter\",\n",
    "    lw=0,\n",
    "    vmax=vmax,\n",
    ")\n",
    "cbar = plt.colorbar(im, cax=cbar_ax)\n",
    "cbar.set_label(\"OR expression\\n(log-normalized)\")\n",
    "cbar.solids.set_rasterized(True)\n",
    "cbar.solids.set_edgecolor(\"face\")\n",
    "cbar.outline.set_visible(False)\n",
    "ax.axis(\"off\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46dcef21-fc76-4220-b098-29e8db79727f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#OR family heatmap\n",
    "df_patient_osn = df_patient[df_patient.osn_clusters.isin((\"iOSN\", \"mOSN\"))].join(\n",
    "    df_olfr\n",
    ")\n",
    "df_family_counts = pd.crosstab(df_patient_osn.family, df_patient_osn.cond)\n",
    "df_pivot_norm = df_family_counts / df_family_counts.sum(0) * 100\n",
    "df_to_plot = df_pivot_norm[[\"normosmic\", \"covid\"]]\n",
    "df_to_plot.columns = [\"control\", \"COVID\"]\n",
    "or_fam_order = natsorted(df_to_plot.index.tolist())\n",
    "df_to_plot = df_to_plot.loc[or_fam_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd0bc605-7bb9-4756-b3c5-18f6430785fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(1, 2))\n",
    "cbar_ax = fig.add_axes([1, 0.2, 0.08, 0.6])\n",
    "sns.heatmap(\n",
    "    df_to_plot,\n",
    "    cmap=\"viridis\",\n",
    "    cbar_kws={\"label\": \"% of OSNs from OR family\"},\n",
    "    cbar_ax=cbar_ax,\n",
    "    ax=ax,\n",
    ")\n",
    "ax.axvline(1, color=\"white\", lw=0.75)\n",
    "ax.set_ylabel(\"OR family\")\n",
    "ax.set_xlabel(None)\n",
    "cbar_ax.xaxis.set_label_position(\"top\")\n",
    "cbar_ax.xaxis.set_ticks_position(\"top\")\n",
    "cbar_ax.xaxis.set_tick_params(pad=0)\n",
    "ax.yaxis.set_tick_params(pad=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "887d8714-4765-490b-8fea-9c423b25122c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculate and plot transcriptomic distance\n",
    "X_pca = adata.obsm[\"X_scVI_OSN\"][is_patient]\n",
    "dmat = metrics.pairwise_distances(X_pca, metric=\"correlation\")\n",
    "df_dmat = pd.DataFrame(\n",
    "    dmat, index=adata.obs_names[is_patient], columns=adata.obs_names[is_patient]\n",
    ")\n",
    "df_corr_nan = df_dmat.copy()\n",
    "df_corr_nan.index.name = \"cell_1\"\n",
    "df_corr_nan.columns.name = \"cell_2\"\n",
    "np.fill_diagonal(df_corr_nan.values, np.nan)\n",
    "df_corr_stack = df_corr_nan.stack().rename(\"corr\").reset_index()\n",
    "df[\"merge_clust\"] = df.osn_clusters.replace({\"GBC\": \"GBC/INP\", \"INP\": \"GBC/INP\"})\n",
    "df_corr_stack = df_corr_stack.merge(\n",
    "    df[[\"cond\", \"merge_clust\"]], left_on=\"cell_1\", right_index=True\n",
    ").merge(\n",
    "    df[[\"cond\", \"merge_clust\"]],\n",
    "    left_on=\"cell_2\",\n",
    "    right_index=True,\n",
    "    suffixes=[\"_1\", \"_2\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "113d6df9-2bd5-422d-81fb-5a9b1b96e026",
   "metadata": {},
   "outputs": [],
   "source": [
    "#median pairwise distance for cell pairs from each cluster\n",
    "df_piv = (\n",
    "    df_corr_stack.groupby([\"cond_1\", \"merge_clust_1\", \"cond_2\", \"merge_clust_2\"])[\n",
    "        \"corr\"\n",
    "    ]\n",
    "    .median()\n",
    "    .reset_index()\n",
    "    .pivot(\n",
    "        index=[\"merge_clust_1\", \"cond_1\"],\n",
    "        columns=[\"merge_clust_2\", \"cond_2\"],\n",
    "        values=\"corr\",\n",
    "    )\n",
    ")\n",
    "new_order = [\"GBC/INP\", \"iOSN\", \"mOSN\"]\n",
    "df_pivot = df_piv[new_order].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a3b0b8b-8b12-4de9-9889-52a3dd13b25e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot\n",
    "o_color = LabelEncoder().fit_transform(df_pivot.index.get_level_values(0))[\n",
    "    :, np.newaxis\n",
    "]\n",
    "m_color = LabelEncoder().fit_transform(df_pivot.index.get_level_values(1))[\n",
    "    np.newaxis, :\n",
    "]\n",
    "deltas = np.where(np.diff(o_color.flatten()) != 0)[0]\n",
    "deltas = np.array([0, *deltas, len(o_color)])\n",
    "width_ratios = [0.1, 1]\n",
    "height_ratios = [0.1, 1]\n",
    "fig = plt.figure(figsize=(sum(width_ratios), sum(height_ratios)))\n",
    "gs = mpl.gridspec.GridSpec(\n",
    "    len(height_ratios),\n",
    "    len(width_ratios),\n",
    "    fig,\n",
    "    0.0,\n",
    "    0.0,\n",
    "    1,\n",
    "    1,\n",
    "    height_ratios=height_ratios,\n",
    "    width_ratios=width_ratios,\n",
    "    wspace=0,\n",
    "    hspace=0,\n",
    ")\n",
    "dist_ax = fig.add_subplot(gs[1, 1], xticks=[], yticks=[], xlabel=None, frameon=False)\n",
    "dist_im = dist_ax.imshow(df_pivot.values, vmin=0, vmax=2, cmap=\"PRGn\")\n",
    "cmap = mpl.colors.LinearSegmentedColormap.from_list(\n",
    "    \"cmap\", (\"#776eb8\", \"#f1ad0c\", \"#666666\")\n",
    ")\n",
    "top_cmap = mpl.colors.LinearSegmentedColormap.from_list(\"cmap\", plt.cm.Set1.colors[:2])\n",
    "\n",
    "left_ax = fig.add_subplot(gs[1, 0], xticks=[], yticks=[], frameon=False)\n",
    "left_ax.imshow(\n",
    "    o_color,\n",
    "    interpolation=\"none\",\n",
    "    cmap=cmap,\n",
    "    aspect=\"auto\",\n",
    "    rasterized=True,\n",
    ")\n",
    "left_ax.set_yticks([0.5, 2.5, 4.5])\n",
    "left_ax.set_yticklabels(new_order)\n",
    "top_ax = fig.add_subplot(gs[0, 1], xticks=[], yticks=[], frameon=False)\n",
    "top_ax.imshow(\n",
    "    m_color,\n",
    "    interpolation=\"none\",\n",
    "    cmap=top_cmap,\n",
    "    aspect=\"auto\",\n",
    "    rasterized=True,\n",
    ")\n",
    "cb_ax = fig.add_axes([1.1, 0.09, 0.07, 0.75])\n",
    "cbar = fig.colorbar(dist_im, cax=cb_ax)\n",
    "cbar.outline.set_visible(False)\n",
    "cbar.set_label(r\"Transcriptome Distance\")\n",
    "cbar.set_ticks(mpl.ticker.MultipleLocator(0.5))\n",
    "legend_TN = [mpl.patches.Patch(color=c, label=l) for l, c in N_COLOR_MAP.items()]\n",
    "leg = top_ax.legend(\n",
    "    handles=legend_TN,\n",
    "    frameon=True,\n",
    "    loc=\"lower center\",\n",
    "    bbox_to_anchor=(0.5, 1),\n",
    "    ncol=2,\n",
    "    columnspacing=0.5,\n",
    "    handletextpad=0.2,\n",
    ")\n",
    "leg.get_frame().set_linewidth(0.5)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jbf18_env",
   "language": "python",
   "name": "jbf18_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
